### PR TITLE
Two improvements to tf.split's shape function

### LINKED ISSUE #19360: tf.split's -1 support doesn't handle zero dimensions
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Colab
- **TensorFlow installed from (source or binary)**: Colab
- **TensorFlow version (use command below)**: ('unknown', '1.7.0') and 1.8.0
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: `tf.split(tf.zeros([0]), [0, -1], axis=-1)`

### Describe the problem
The variable size version of `tf.split` (`SplitV` in C++) allows one of the sizes to be -1.  The corresponding output will expand as necessary so that the total output size matches the input.

Unfortunately, the -1 support currently assumes the -1 dimension corresponds to **positive** size.  It should handle zero as well.  E.g., this should work, but it doesn't:

```
>>> tf.split(tf.zeros([0]), [0, -1], axis=-1)
Traceback (most recent call last):
  File "/Users/irving/anaconda/envs/openai/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1592, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Sum of output sizes must match the size of the original Tensor along the split dimension or the sum of the positive sizes must be less if it contains a -1 for 'split_3' (op: 'SplitV') with input shapes: [0], [2], [] and with computed input tensors: input[1] = <0 -1>, input[2] = <-1>.
```

By comparison, the positive case works fine:

```
>>> tf.split(tf.zeros([1]), [0, -1], axis=-1)
[<tf.Tensor 'split_4:0' shape=(0,) dtype=float32>, <tf.Tensor 'split_4:1' shape=(?,) dtype=float32>]
```

### PR DESCRIPTION
**Pull request by girving**
1. Don't break if the size corresponding to a -1 is 0.
2. Infer the size corresponding to a -1 if possible.

Fixes #19360.

### CODE REVIEW

**Comment by caisq** [comment]
@girving I can take a look at this later today.

**Review by jhseu — DISMISSED** [review]

File: tensorflow/python/kernel_tests/split_op_test.py [hunk 1]
```diff
@@ -153,6 +153,14 @@ def _testSpecialCasesVariable(self):
       self.assertAllEqual(result[0], inp[0:1, :])
       self.assertAllEqual(result[1], inp[1:4, :])
 
+      result = self.evaluate(array_ops.split(inp, [-1, 4], 0))
+      self.assertAllEqual(result[0], inp[0:0, :])
+      self.assertAllEqual(result[1], inp[0:4, :])
+
+      result = self.evaluate(array_ops.split(inp, [4, -1], 0))
```
Comment (caisq) [inline tensorflow/python/kernel_tests/split_op_test.py]: Maybe add a test to cover splitting with -1's in a non-first axis as well.

**Review by caisq — COMMENTED** [review]
Thanks for the PR @girving. I know @jhseu already approved it. But I wrote a couple of comments and didn't get a chance to finish the review before. I just published the minor comments. Feel free to address or ignore them.

File: tensorflow/core/ops/array_ops.cc [hunk 2]
```diff
@@ -631,38 +631,39 @@ REGISTER_OP("SplitV")
           return errors::InvalidArgument(
               "Length of size_splits should be equal to num_outputs");
         }
-        int64_t cumsum_outputs = 0;
+        int64_t total_size = 0;
         bool has_neg_one = false;
+        for (const auto size : data) {
+          if (size == -1) {
+            if (has_neg_one)
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/core/ops/array_ops.cc]: Resolved.

File: tensorflow/python/kernel_tests/split_op_test.py [hunk 3]
```diff
@@ -153,6 +153,14 @@ def _testSpecialCasesVariable(self):
       self.assertAllEqual(result[0], inp[0:1, :])
       self.assertAllEqual(result[1], inp[1:4, :])
 
+      result = self.evaluate(array_ops.split(inp, [-1, 4], 0))
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/python/kernel_tests/split_op_test.py]: Done.

File: tensorflow/python/kernel_tests/split_op_test.py (hunk 1 above)
Comment (girving) [reply tensorflow/python/kernel_tests/split_op_test.py]: Done.

**Comment by drpngx** [comment]
Actually @girving could you push again?

**Comment by girving** [comment]
@drpngx Not before I finish running the tests.

**Comment by girving** [comment]
Pushed.  Apologies for delay.  (Cc @drpngx)

**Comment by girving** [comment]
One more commit to (hopefully) fix clang format errors.

**Comment by yifeif** [comment]
Thanks @girving! Looks like clang-format passed, but the build status reporting is broken. Let me fix that on the infra side.

**Pull request merged**